File: BalancedBinaryTree/dfs.java
public class Solution {
    private boolean balanced = true;
    public boolean isBalanced(TreeNode root) {
        findDepth(root, 1);
        return balanced;
    }

   private int findDepth(TreeNode node, int level){
        if (!balanced || node == null){
            return level-1;
        }
        int depthL = findDepth(node.left, level+1);
        int depthR = findDepth(node.right, level+1);
        int max = Math.max(depthL, depthR);
        int min = Math.min(depthL, depthR);
        if (max-min > 1){
            balanced = false;
        }
        return max;
    }
}

File: BalancedBinaryTree/dfs.cpp
class Solution {
    int balanceness(TreeNode *root) { // return depth if balanced, else -1
        if(!root) return 0;
        int ld=balanceness(root->left);
        if(ld>=0) {
            int rd=balanceness(root->right);
            if(rd>=0 && std::abs(ld-rd)<2) return 1+std::max(ld,rd);
        }
        return -1;
    }
public:
    bool isBalanced(TreeNode* root) {
        return balanceness(root)>=0;
    }
};
